hack/e2e.go

```diff
@@ -332,19 +347,217 @@ func Build() error {
 	return nil
 }
 
-func TearDown() error {
-	return finishRunning("teardown", exec.Command("./hack/e2e-internal/e2e-down.sh"))
+type deployer interface {
+	Up() error
+	IsUp() bool
+	SetupKubecfg() error
+	Down() error
+	Destroy()
+}
+
+func getDeployer() (deployer, error) {
+	switch *deployment {
+	case "bash":
+		return bash{}, nil
+	case "kops":
+		return NewKops()
+	default:
+		return nil, fmt.Errorf("Unknown deployment strategy %q", *deployment)
+	}
 }
 
-func Up() error {
+type bash struct{}
+
+func (b bash) Up() error {
 	return finishRunning("up", exec.Command("./hack/e2e-internal/e2e-up.sh"))
 }
 
-// Is the e2e cluster up?
-func IsUp() bool {
+func (b bash) IsUp() bool {
 	return finishRunning("get status", exec.Command("./hack/e2e-internal/e2e-status.sh")) == nil
 }
 
+func (b bash) SetupKubecfg() error {
+	return nil
+}
+
+func (b bash) Down() error {
+	return finishRunning("teardown", exec.Command("./hack/e2e-internal/e2e-down.sh"))
+}
+
+func (b bash) Destroy() {}
+
+type kops struct {
+	binary    string
+	tmpBinary bool
+	zones     []string
+	nodes     int
+	cluster   string
+	kubecfg   string
+}
+
+func NewKops() (*kops, error) {
+	if os.Getenv("KOPS_STATE_STORE") == "" {
+		return nil, fmt.Errorf("KOPS_STATE_STORE must be set to a valid S3 path for kops deployment")
+	}
+	// Presume the kops binary was supplied to us.
+	binary := os.Getenv("KOPS_BINARY")
```

Comment:
Talked offline, and then I worked some more with this. I think I can do the download entirely in shell in `.yaml` pretty easily, so I'm not sure it's worth the separate binary. I reworked it to be all flags and take the necessary bits coming in, and I can iterate further from there.